core/domain/exp_services.py

```diff
@@ -2053,6 +2053,245 @@ def get_interaction_id_for_state(exp_id, state_name):
+    """
+    # Fetch the model associated with the unique_progress_url_id.
+    model = exp_models.TransientCheckpointUrlModel.get(unique_progress_url_id)
+
+    # Create a model if it doesn't already exist.
+    if model is None:
+        model = exp_models.TransientCheckpointUrlModel.create(
+            exploration_id, unique_progress_url_id)
+
+    current_exploration = exp_fetchers.get_exploration_by_id(
+        exploration_id, True, exp_version)
+
+    # If the exploration is being visited the first time.
+    if model.furthest_reached_checkpoint_state_name is None:
+        model.furthest_reached_checkpoint_exp_version = exp_version
+        model.furthest_reached_checkpoint_state_name = state_name
+    elif model.furthest_reached_checkpoint_exp_version < exp_version:
+        furthest_reached_checkpoint_exp = (
+            exp_fetchers.get_exploration_by_id(
+                exploration_id,
+                strict=True,
+                version=model.furthest_reached_checkpoint_exp_version
+            )
+        )
+        checkpoints_in_current_exp = _get_checkpoints_in_order(
+            current_exploration.init_state_name, current_exploration.states)
+        checkpoints_in_older_exp = _get_checkpoints_in_order(
+            furthest_reached_checkpoint_exp.init_state_name,
+            furthest_reached_checkpoint_exp.states)
+
+        # Get the furthest reached checkpoint in current exploration.
+        furthest_reached_checkpoint_in_current_exp = (
+            _get_most_distant_reached_checkpoint_in_current_exploration(
+                checkpoints_in_current_exp,
+                checkpoints_in_older_exp,
+                model.furthest_reached_checkpoint_state_name
+            )
+        )
+
+        # If the furthest reached checkpoint doesn't exist in current
+        # exploration.
+        if furthest_reached_checkpoint_in_current_exp is None:
+            model.furthest_reached_checkpoint_exp_version = (
+                exp_version)
+            model.furthest_reached_checkpoint_state_name = state_name
+        else:
+            # Index of the furthest reached checkpoint.
+            frc_index = checkpoints_in_current_exp.index(
+                furthest_reached_checkpoint_in_current_exp)
+            # If furthest reached checkpoint is behind most recently
+            # reached checkpoint.
+            if frc_index <= checkpoints_in_current_exp.index(state_name):
+                model.furthest_reached_checkpoint_exp_version = (
+                    exp_version)
+                model.furthest_reached_checkpoint_state_name = (
+                    state_name)
+
+    model.most_recently_reached_checkpoint_exp_version = exp_version
+    model.most_recently_reached_checkpoint_state_name = state_name
+    model.creation_timestamp = datetime.datetime.utcnow()
+    model.update_timestamps()
+    model.put()
+
+
+def sync_learner_checkpoint_progress_with_current_exp_version(
```

Comment:
Ditto, can you make sure to reuse the code that is already present and avoid writing same code multiple times.

Reply:
This one's actually a different function.

Reply:
The only difference that I see is the the first line of getting the model and the last return statement, why not combine these 2 functions and add conditional handling of these 2 lines for logged in and logged out experience. This will ensure 3 things:
1. Code re-usability
2. The way we handle getting the latest progress remain consistent for both logged in and out users in the future as we will only have to change the things at one place
3. No confusion occurs between the functions as both of them have the same names

Reply:
Actually, I think it's better to keep them as separate functions because:
1. Merging them into a single function would mean one extra argument -- unique_progress_url_id or user_id has to be passed around None. 
2. Would ensure more readable and specific code.
3. I'll change the names to ensure zero confusion.

What do you think?